Lib/subprocess.py

```diff
@@ -1414,6 +1419,19 @@ def _execute_child(self, args, executable, preexec_fn, close_fds,
 
             if executable is None:
                 executable = args[0]
+
+            if (preexec_fn is None
+                    and not close_fds
+                    and not pass_fds
+                    and cwd is None
+                    and env is None
+                    and not restore_signals
+                    and not start_new_session
+                    and os.path.dirname(executable)
+                    and self.stdin is self.stdout is self.stderr is None):
+                self._posix_spawn(args, executable)
```

Comment:
Checking `self.std*` is still not enough. Those fields are for the parent's ends of pipes, but if a caller redirects to a descriptor, there is no such end. For example, `subprocess.call(['/bin/sleep', '1000'], stdout=fd, close_fds=False, restore_signals=False)` doesn't work properly.

Reply:
@izbyshev: Oh right! I fixed the test:  test  (p2cread, p2cwrite, c2pread, c2pwrite, errread, errwrite) instead of (self.stdin, self.stdout, self.stderr). Is it better now?

Reply:
Yes, LGTM now.

There is another potential problem with `posix_spawn`: [asynchronous error notification](http://pubs.opengroup.org/onlinepubs/9699919799/functions/posix_spawn.html). `_posixsubprocess` uses a pipe to notify the parent about all errors up to and including `execve` errors, so its style is synchronous. Glibc and musl use the pipe-based notification too, but they are not required to do that, and I don't know what other implementations do.